### PR TITLE
Fix: cluster-servant: check for corosync 2Node mode

### PR DESCRIPTION
**Pull request by wenningerk**
Quick and simple approach for tackling the issue that sbd won't suicide when loosing disk & 2nd node on a 2Node cluster.

### CODE REVIEW

**Comment by wenningerk** [comment]
now enhanced to track quorum.two_node

File: src/sbd-cluster.c [hunk 1]
```diff
@@ -67,21 +70,130 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+    }
+    if (track_handle) {
+        cmap_track_delete(cmap_handle, track_handle);
+    }
+    if (cmap_handle) {
+        cmap_finalize(cmap_handle);
+    }
+    track_handle = 0;
+    cmap_handle = 0;
+    cmap_source = NULL;
```
Comment (jnpkrn) [inline src/sbd-cluster.c]: Depends on the taste, but would zero the variables along with respective destructors, for more clarity/less redundancy.

File: src/sbd-cluster.c [hunk 2]
```diff
@@ -67,21 +70,130 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+        if (!two_node || !ever_seen_both || cpg_membership_entries > 1) {
+            set_servant_health(pcmk_health_online, LOG_INFO,
+                           "Connected to %s (%u members)",
+                           name_for_cluster_type(get_cluster_type()),
+                           cpg_membership_entries
+                          );
+        } else {
+            /* should ask votequorum but like this races between
+             * votequorum & membership callbacks are no issue
+             * and no bindings to votequorum api are needed
```
Comment (jnpkrn) [inline src/sbd-cluster.c]: ENOTPARSING

File: src/sbd-cluster.c [hunk 3]
```diff
@@ -67,21 +70,137 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+    uint8_t two_node_u8 = 0;
+    int cmap_fd;
+
+    if (!track_handle) {
+        if (cmap_initialize(&cmap_handle) != CS_OK) {
+            cl_log(LOG_WARNING, "Cannot initialize CMAP service\n");
+            goto out;
+        }
+
+        if (cmap_track_add(cmap_handle, "quorum.two_node", CMAP_TRACK_MODIFY,
```
Comment (jfriesse) [inline src/sbd-cluster.c]: Tracking just modify operation is not enough. You have to track also add and delete operations (CMAP_TRACK_ADD and CMAP_TRACK_DELETE)

File: src/sbd-cluster.c [hunk 4]
```diff
@@ -67,21 +70,137 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+static void sbd_cmap_notify_fn(
+    cmap_handle_t cmap_handle,
+    cmap_track_handle_t cmap_track_handle,
+    int32_t event,
+    const char *key_name,
+    struct cmap_notify_value new_val,
+    struct cmap_notify_value old_val,
+    void *user_data)
+{
+    if ((event == CMAP_TRACK_MODIFY) &&
```
Comment (jfriesse) [inline src/sbd-cluster.c]: Tracking just modify operation is not enough. You have to track also add and delete operations (CMAP_TRACK_ADD and CMAP_TRACK_DELETE)

Comment (wenningerk) [inline src/sbd-cluster.c]: Whether I had 2Node enabled or disabled on corosync-startup quorum.two_node was always available in cmap and it is definitely working. Which cases would need add & delete as well?
Or just for completeness in general?

Comment (jfriesse) [inline src/sbd-cluster.c]: Explained via IRC.

File: src/sbd-cluster.c [hunk 5]
```diff
@@ -67,21 +70,149 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+            set_servant_health(pcmk_health_online, LOG_INFO,
+                           "Connected to %s (%u members)",
+                           name_for_cluster_type(get_cluster_type()),
+                           cpg_membership_entries
+                          );
+        } else {
+            /* Alternative would be asking votequorum for number of votes.
+             * Using pacemaker's cpg as source for number of active nodes
+             * avoids binding to an additional library, is definitely
+             * less code to write and we wouldn't have to combina data
```
Comment (jnpkrn) [inline src/sbd-cluster.c]: combine

File: src/sbd-cluster.c [hunk 6]
```diff
@@ -67,21 +70,149 @@ sbd_plugin_membership_dispatch(cpg_handle_t handle,
...
+static cmap_handle_t cmap_handle = 0;
+static cmap_track_handle_t track_handle = 0;
+static int cpg_membership_entries = -1;
+static GSource *cmap_source = NULL;
+
+void
+sbd_cpg_membership_health_update()
+{
+    if(cpg_membership_entries > 0) {
+        if (!two_node || !ever_seen_both || cpg_membership_entries > 1) {
```
Comment (atomictoaster) [inline src/sbd-cluster.c]: Something seems off here.

Either `ever_seen_both` is badly named ('both' implies 2 nodes) or perhaps `!two_node` should be just `two_node` here.  Or possibly both, I'm really not sure what you're going for here.  

It seems like a 5 node cluster with 1 active node would end up setting `pcmk_health_online`, even if it had previously been able to see everyone.  Which seems wrong.  Maybe you want s/||/&&/ ?

Might be a good idea to include a comment describing the scenario you're trying to detect.

Comment (beekhof) [inline src/sbd-cluster.c]: Something seems off here. Either `ever_seen_both` is badly named ('both' implies 2 nodes) or perhaps `!two_node` should be just `two_node` here. Or possibly both, I'm really not sure what you're going for here.

Based on this logic, the sole survivor of a cluster with 5 nodes would end up setting `pcmk_health_online` - which doesn't sound right.  Maybe you mean s/||/&&/ ?

Either way, I suggest a nice big comment in the code to explain the case you're trying to deal with.

Comment (wenningerk) [inline src/sbd-cluster.c]: I'll would add a comment like that:
Caveat: This is not to replace the quorum-handling in pacemaker-servant.
It is just there to help out in the one and only case where 2Node-flag makes quorum coming from pengine via cib unusable for pacemaker-servant. Reason why it is not implemented there is that cluster-servant already interacts with corosync and pacemaker-servant just reads the cib and derives everything from there.
If !two_node or !ever_seen_both it falls back to the previous behavior.
ever_seen_both may be named badly but it is there to assure startup-cases and it is set once more than 1 node (most likely this node himself) has been seen.

Comment (kgaillot) [inline src/sbd-cluster.c]: Maybe more readable:

quorum_is_suspect = (two_node && ever_seen_both && cpg_membership_entries == 1)
if (!quorum_is_suspect)

File: configure.ac [hunk 7]
```diff
@@ -42,8 +44,10 @@ PKG_CHECK_MODULES(libqb, [libqb])
 CPPFLAGS="$CPPFLAGS -Werror"
 if test $HAVE_pacemaker = 0 -a $HAVE_pcmk = 0; then
     AC_MSG_ERROR(No package 'pacemaker' found)
+elif test $HAVE_cmap = 0; then
+    AC_MSG_ERROR(No package 'cmap' found)
```
Comment (kgaillot) [inline configure.ac]: I don't think corosync 1 has libcmap. sbd supports corosync 1 + pacemaker plugin. We don't have to implement the new behavior for the old stack, but we shouldn't break it

**Comment by wenningerk** [comment]
Should now as well build if cmap isn't available - just no 2-node-checking then.
Took the hint from Ken to introduce quorum_is_suspect to make it clearer why we are doing all this.

**Comment by kgaillot** [comment]
Looks good to merge to me

BTW we should set up travis for sbd, we can at least verify it can compile and be packaged

**Comment by wenningerk** [comment]
yes - definitely for building and packaging at least.
already thought of what we could do more but unfortunately there is nothing as easy as cib in transition out here.

**Pull request merged**